### ISSUE TITLE
[BUG] cloneset scale choose pod panic when diff > len(pods)

### ISSUE THREAD

**Issue by shiyan2016** [openkruise/kruise#306]
**What happened**:
scale choose pod panic when diff > len(pods)

cloneset status: 
```bash
NAME                  DESIRED   UPDATED   UPDATED_READY   READY   TOTAL   AGE
default               4         0         0               5       5       2d2h
```
panic log:
```bash
I0702 10:23:05.325465       1 cloneset_scale.go:100] CloneSet default/test begin to scale in 1 pods including 0 (current rev)
I0702 10:23:05.325500       1 cloneset_controller.go:175] Finished syncing CloneSet default/test, cost 105.113125ms
E0702 10:23:05.325666       1 runtime.go:69] Observed a panic: runtime.boundsError{x:1, y:0, signed:true, code:0x2} (runtime error: slice bounds out of range [:1] with capacity 0)
/workspace/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/workspace/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/workspace/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:679
/usr/local/go/src/runtime/panic.go:93
/workspace/pkg/controller/cloneset/scale/cloneset_scale_util.go:104
/workspace/pkg/controller/cloneset/scale/cloneset_scale_util.go:114
/workspace/pkg/controller/cloneset/scale/cloneset_scale.go:103
/workspace/pkg/controller/cloneset/cloneset_controller.go:302
/workspace/pkg/controller/cloneset/cloneset_controller.go:254
/workspace/pkg/controller/cloneset/cloneset_controller.go:165
/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1357
panic: runtime error: slice bounds out of range [:1] with capacity 0 [recovered]
        panic: runtime error: slice bounds out of range [:1] with capacity 0
```
**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
1. create an cloneset with replicas=5 
2. update the cloneset's pod template and partition to 5
3. scale the clonset to replicas=4

**Anything else we need to know?**:

**Environment**:
- Kruise version: master
- Kubernetes version (use `kubectl version`): v1.17.3

**Comment by FillZpp** [comment]
@shiyan2016 Thanks for reporting this. Could you describe how to reproduce it?

**Comment by shiyan2016** [comment]
@FillZpp I have update how to reproduce it. In this case, it will panic. And we should discuss another question, when partition > replicas, how many old revision pods should we keep, paritition or replicas? If keep partition, we do not need update the calculateDiffs func. If keep replicas, then we should update the func calulate currentRevDiff  as belows:
```
currentRevDiff = notUpdatedPods - integer.IntMin(int(*cs.Spec.UpdateStrategy.Partition), int(*cs.Spec.Replicas))
```

**Comment by FillZpp** [comment]
@shiyan2016 Get it. I think it is a bug in `calculateDiffs`, not `choosePodsToDelete `. We should keep replicas during partition > replicas and fix the func as you said.

Additional unit tests in `TestCalculateDiffs` could be better:
```golang
		{
			name: "scale in without revConsistent 6",
			cs: &appsv1alpha1.CloneSet{
				Spec: appsv1alpha1.CloneSetSpec{
					Replicas: utilpointer.Int32Ptr(9),
					UpdateStrategy: appsv1alpha1.CloneSetUpdateStrategy{
						Partition: utilpointer.Int32Ptr(10),
					},
				},
			},
			revConsistent:          false,
			totalPods:              10,
			notUpdatedPods:         10,
			expectedTotalDiff:      1,
			expectedCurrentRevDiff: 1,
		},
		{
			name: "scale in without revConsistent 7",
			cs: &appsv1alpha1.CloneSet{
				Spec: appsv1alpha1.CloneSetSpec{
					Replicas: utilpointer.Int32Ptr(9),
					UpdateStrategy: appsv1alpha1.CloneSetUpdateStrategy{
						MaxSurge:  &intOrStr1,
						Partition: utilpointer.Int32Ptr(10),
					},
				},
			},
			revConsistent:          false,
			totalPods:              10,
			notUpdatedPods:         10,
			expectedTotalDiff:      0,
			expectedCurrentRevDiff: 1,
		},
```

**Comment by shiyan2016** [comment]
@FillZpp ok, I will submit another PR. And for the choosePodsToDelete func I think it is better to handle like https://github.com/openkruise/kruise/pull/305/files and we can add a warnning log to remind us coming diff > len(pods).

**Comment by FillZpp** [comment]
@shiyan2016 Great, I have added a comment in #305 . Thanks for your contributing!

**Comment by shiyan2016** [comment]
@FillZpp  I submit another PR #309 , maybe you can closed the PR #305 and review the PR #309. Thanks.
